compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -188,6 +326,8 @@ impl<'a> Parser<'a> {
             RecoverReturnSign::Yes,
             None,
             RecoverQuestionMark::No,
+            // Disable the entry of parsing `$ty:ty` in macros
+            AllowAnonStructOrUnion::No,
```

Comment:
`$ty:ty` is the only reason why `struct/union { ... }`s are parsed as general types, rather than a part of the field definition syntax (only `struct FieldDef` is extended with extra possibilities, not `enum TyKind`).

If `struct { foo: u8 }` cannot be passed to where `ty` is expected, then most of the complexity introduced by this PR stops making sense.